src/vertex_tools.rs

```diff
@@ -0,0 +1,1043 @@
+    ) {
+        let face_points = (
+            &Point3::new(0.0, 1.0, 0.0),
+            &Point3::new(-0.866025, -0.5, 0.0),
+            &Point3::new(0.866025, -0.5, 0.0),
+        );
+
+        let ray_origin = Point3::new(0.0, 0.0, 0.0);
+        let ray_vector = Vector3::new(0.0, 0.0, 0.0);
+
+        let point_on_triangle_calculated =
+            ray_intersects_triangle(&ray_origin, &ray_vector, face_points)
+                .expect("Point is not on the triangle.");
+
+        assert_eq!(point_on_triangle_calculated.point, ray_origin);
+        assert_eq!(point_on_triangle_calculated.distance, 0.0);
+    }
+
+    #[test]
+    fn test_vertex_tools_ray_intersects_triangle_for_horizontal_triangle_returns_point_inside() {
+        let face_points = (
+            &Point3::new(0.0, 1.0, 0.0),
+            &Point3::new(-0.866025, -0.5, 0.0),
+            &Point3::new(0.866025, -0.5, 0.0),
+        );
+
+        let ray_origin = Point3::new(0.25, 0.25, 0.25);
+        let ray_vector = Vector3::new(0.0, 0.0, -1.0);
+
+        let point_on_triangle_correct = Point3::new(0.25, 0.25, 0.0);
+        let distance_correct = nalgebra::distance(&ray_origin, &point_on_triangle_correct);
+
+        let point_on_triangle_calculated =
+            ray_intersects_triangle(&ray_origin, &ray_vector, face_points)
+                .expect("Point is not on the triangle.");
+
+        assert_eq!(
+            point_on_triangle_calculated.point,
+            point_on_triangle_correct
+        );
+        assert_eq!(point_on_triangle_calculated.distance, distance_correct);
+    }
+
+    #[test]
+    fn test_vertex_tools_ray_intersects_triangle_for_horizontal_triangle_returns_none_because_ray_misses(
+    ) {
+        let face_points = (
+            &Point3::new(0.0, 1.0, 0.0),
+            &Point3::new(-0.866025, -0.5, 0.0),
+            &Point3::new(0.866025, -0.5, 0.0),
+        );
+
+        let ray_origin = Point3::new(1.25, 0.25, 0.25);
+        let ray_vector = Vector3::new(0.0, 0.0, -1.0);
+
+        let point_on_triangle_calculated =
+            ray_intersects_triangle(&ray_origin, &ray_vector, face_points);
+
+        // There is no intersection of the triangle and ray
+        assert_eq!(None, point_on_triangle_calculated);
+    }
+
+    #[test]
+    fn test_vertex_tools_ray_intersects_triangle_for_horizontal_triangle_returns_none_because_ray_parallel(
+    ) {
+        let face_points = (
+            &Point3::new(0.0, 1.0, 0.0),
+            &Point3::new(-0.866025, -0.5, 0.0),
+            &Point3::new(0.866025, -0.5, 0.0),
+        );
+
+        let ray_origin = Point3::new(0.25, 0.25, 0.25);
+        let ray_vector = Vector3::new(1.0, 0.0, 0.0);
+
+        let point_on_triangle_calculated =
+            ray_intersects_triangle(&ray_origin, &ray_vector, face_points);
+
+        // There is no intersection of the triangle and ray
+        assert_eq!(None, point_on_triangle_calculated);
+    }
```

Comment:
Aren't the previous 2 tests identical in their meaning? Or is the reason of miss significant?

Reply:
Yes, it’s a different case. When the ray misses the triangle, it still hits the triangle plane. If there was a mistake in the logic, this might report as a hit.

Reply:
Yes, it’s a different case. When the ray misses the triangle, it still hits the triangle plane. If there was a mistake in the logic, this might report as a hit.